File: linked list/LinkedList.cpp
#include "LinkedList.h"
using namespace std;

LinkedList::LinkedList()
{
    HEAD=NULL;
    TAIL=NULL;
}

LinkedList::~LinkedList(){}

void LinkedList::addToHead(int data)
{
    Node *newNode=new Node();
    newNode->info=data;
    newNode->next=HEAD;
    HEAD=newNode;
    if (TAIL==NULL)
    {
        TAIL=newNode;
    }
}

void LinkedList::addToTail(int data)
{
    Node *newNode=new Node();
    newNode->info=data;
    TAIL->next=newNode;
    TAIL=newNode;
    if (HEAD==NULL)
    {
        HEAD=newNode;
    }
}

void LinkedList::traverse()
{
    Node *temp=HEAD;
    while (temp!=NULL)
    {
        cout<<temp->info<<endl;
        temp=temp->next;
    }
}

bool LinkedList::search(int data)
{
    Node *temp=HEAD;
    while (temp!=NULL)
    {
        if (temp->info==data)
        {
            return true;
            break;
        }
        temp=temp->next;
    }
    return false;
}

Node* LinkedList::retreive(int data)
{
    Node* temp=HEAD;
    while(temp!=NULL)
    {
        if (temp->info==data)
        {
            return temp;
            break;
        }
        temp=temp->next;
    }
    cout<<"Data not present"<<endl;
}

void LinkedList::add(int data,Node *predecessor)
{
    Node *newNode=new Node();
    newNode->info=data;
    newNode->next=predecessor->next;
    predecessor->next=newNode;
}

File: Python/linked_list/linked_list.py
class Node:
    def __init__(self, value):
        self.data = value
        self.next = None


class LinkedList:
    def __init__(self):
        self.HEAD = None
        self.TAIL = None

    def add_to_head(self, data):
        new_node = Node(data)
        if not self.HEAD:
            self.HEAD = new_node
        else:
            new_node.next = self.HEAD
            self.HEAD = new_node
        if not self.TAIL:
            self.TAIL = new_node
    
    def add_to_tail(self, data):
        new_node = Node(data)
        if not self.HEAD:
            self.HEAD = new_node
        if not self.TAIL:
            self.TAIL = new_node
        else:
            self.TAIL.next = new_node
            self.TAIL = new_node
    
    def delete_from_head(self):
        node = self.HEAD
        self.HEAD = node.next
        del node
    
    def delete_from_tail(self):
        node = self.HEAD
        while node.next:
            prev = node
            node = node.next
        del self.TAIL
        self.TAIL = prev
        self.TAIL.next = None
    
    def search(self, data):
        node = self.HEAD
        while node:
            if node.data == data:
                return True
            node = node.next
        return False
    
    def traverse(self):
        node = self.HEAD
        values = []
        while node:
            values.append(str(node.data))
            node = node.next
        return " -> ".join(values) if values else "List Empty!"